File: examples/Erlang/taskwork2.erl
%%
%% Task worker
%% Connects to ventilator on tcp://localhost:5557, sink on tcp://localhost:5558 and subscribe to tcp://localhost:5559
%% Receives task and push result to sink. Terminate when receives <<"KILL">> message from sink
%% Mikhail Kulemin <mihkulemin@gmail.com>
%%

-module(taskwork2).
-export([run/0]).

run() ->
	{ok,Context} = erlzmq:context(),
	
	{ok,Receiver} = erlzmq:socket(Context,[pull,{active,true}]),
	ok = erlzmq:connect(Receiver,"tcp://localhost:5557"),

	{ok,Sender} = erlzmq:socket(Context,push),
	ok = erlzmq:connect(Sender,"tcp://localhost:5558"),

	{ok,Controller} = erlzmq:socket(Context,[sub,{active,true}]),
	ok = erlzmq:connect(Controller,"tcp://localhost:5559"),
	ok = erlzmq:setsockopt(Controller,subscribe,""),



	ok = loop(Receiver,Sender,Controller),

	ok = erlzmq:close(Receiver),
	ok = erlzmq:close(Sender),
	ok = erlzmq:close(Controller),
	ok = erlzmq:term(Context).

loop(Receiver,Sender,Controller) ->
	receive 
		{zmq,Receiver,Work} ->
			io:format(".",[]),

			Time = list_to_integer(binary_to_list(Work)),
			timer:sleep(Time),

			ok = erlzmq:send(Sender,<<"">>),
			loop(Receiver,Sender,Controller);
		{zmq,Controller,<<"KILL">>} ->
			ok
	end.	
	

File: examples/Haskell/taskwork2.hs
{-# LANGUAGE OverloadedStrings #-}

--  Task worker - design 2
--  Adds pub-sub flow to receive and respond to kill signal

module Main where

import           Control.Concurrent
import           Control.Monad
import qualified Data.ByteString.Char8 as BS
import           Data.Function
import           System.IO
import           System.ZMQ4.Monadic
import           Text.Printf

main :: IO ()
main = runZMQ $ do
    -- Socket to receive messages on
    receiver <- socket Pull
    connect receiver "tcp://localhost:5557"

    -- Socket to send messages to
    sender <- socket Push
    connect sender "tcp://localhost:5558"

    controller <- socket Sub
    connect controller "tcp://localhost:5559"
    subscribe controller ""

    liftIO $ hSetBuffering stdout NoBuffering

    fix $ \loop -> do
        [receiver_events, controller_events] <-
            poll (-1) [ Sock receiver   [In] Nothing
                      , Sock controller [In] Nothing
                      ]

        when (receiver_events /= []) $ do
            string <- BS.unpack <$> receive receiver
            liftIO $ printf "%s." string              -- Show the progress
            liftIO $ threadDelay (read string * 1000) -- Do the work
            send sender [] ""                         -- Send results to sink

        -- Any waiting controller command acts as 'KILL'
        unless (controller_events /= [])
            loop
